File: fetch_issues.cpp
#include <iostream>
#include <string>
#include <nlohmann/json.hpp>
#include <cpr/cpr.h>
#include <fstream>
#include <vector>
#include <htmlcxx/html/ParserDom.h>

// configurations
const std::string GITHUB_REPO = "just-koala/just-koala.github.io";
const std::string GITHUB_TOKEN = getenv("KOALA_TOKEN");

nlohmann::json fetch_issues(const std::string &repo)
{
    std::string url = "https://api.github.com/repos/" + repo + "/issues";
    cpr::Response r = cpr::Get(cpr::Url{url}, cpr::Header{{"Authorization", "token " + GITHUB_TOKEN}});
    return nlohmann::json::parse(r.text);
}

// Generate blog posts from issues
std::string generate_post(const nlohmann::json &issue)
{
    std::ifstream template_file("template.html");
    std::string template_str((std::istreambuf_iterator<char>(template_file)), std::istreambuf_iterator<char>());

    std::string title = issue["title"];
    std::string body = issue["body"];

    size_t title_pos = template_str.find("{{ title }}");
    if (title_pos != std::string::npos)
    {
        template_str.replace(title_pos, 10, title);
    }

    size_t content_pos = template_str.find("{{ content }}");
    if (content_pos != std::string::npos)
    {
        template_str.replace(content_pos, 12, body);
    }

    return template_str;
}

int main()
{
    nlohmann::json issues = fetch_issues(GITHUB_REPO);
    std::ofstream posts_dir("posts");

    for (auto &issue : issues)
    {
        for (auto &label : issue["labels"])
        {
            if (label["name"] == "blog")
            {
                std::string filename = "posts/" + issue["title"].get<std::string>() + ".html";
                std::ofstream post_file(filename);
                post_file << generate_post(issue);
            }
        }
    }
    return 0;
}

File: fetch_issues.mjs
import { Octokit } from "@octokit/rest";
import fs from 'fs';
import path from 'path';

// Configuration
const GITHUB_TOKEN = process.env.MY_GITHUB_TOKEN;
const GITHUB_REPO = 'just-koala/just-koala.github.io';
const [owner, repo] = GITHUB_REPO.split('/');

// Initialize Octokit
const octokit = new Octokit({
  auth: GITHUB_TOKEN
});

// Fetch issues with the label 'blogpost'
async function fetchIssues() {
  const { data: issues } = await octokit.issues.listForRepo({
    owner,
    repo,
    labels: 'blogpost',
    state: 'open'
  });
  return issues;
}

// Generate HTML from issue data
function generateHtml(issue) {
    const title = issue.title;
    const body = issue.body;
  
    const html = `
    <!DOCTYPE html>
    <html lang="en">
    <head>
      <meta charset="UTF-8">
      <meta name="viewport" content="width=device-width, initial-scale=1.0">
      <title>${title}</title>
    </head>
    <body>
      <h1>${title}</h1>
      <div>${body}</div>
    </body>
    </html>
    `;
    return html;
  }

// Main function
(async () => {
    const issues = await fetchIssues();
    if (!fs.existsSync('posts')) {
      fs.mkdirSync('posts');
    }
    for (const issue of issues) {
      const filename = `posts/${issue.title.replace(/\s+/g, '_')}.html`;
      const html = generateHtml(issue);
      fs.writeFileSync(filename, html);
      console.log(`Generated ${filename}`);
    }
  })();
